core/trie/trie.go

```diff
@@ -52,7 +52,7 @@ func NewTrie(storage Storage, height uint, rootKey *bitset.BitSet) *Trie {
 
 // RunOnTempTrie creates an in-memory Trie of height `height` and runs `do` on that Trie
 func RunOnTempTrie(height uint, do func(*Trie) error) error {
-	db, err := db.NewInMemoryDb()
+	db, err := pebble.NewMem()
```

Comment:
We don't, I didn't like the way it was before either since db package contained pebble and badger implementation albeit private. The issue is due to the convenience of `RunOnTempTrie` which is causing us to import pebble. If we were to pass it as an argument then it would defeat the purpose of having `RunOnTempTrie`. Also, it goes back to the conversation we had regarding where the interface should be kept and whether we should redeclare interfaces with a slight difference in every package which requires a DB. 

I will think of a way to get rid of this reference.

Reply:
How does this sound;

a package named `kvdb` that defines the interface, a `pebble` package that imlements the interface and a `db` package that exposes `NewDb` like instantiation functions. Test could go in to either `kvdb` or `db`

Reply:
Yeah, we could do something like that. However, I don't like the idea of having a package solely for providing default `NewDB` functions it hides from the user what they are using.

I have made the following changes:

- Reintroduce memory storage which is what `RunOnTempTrie` now operates on. This removes any references to pebble in core (except test functions which is ok)
- Trie package now solely defines what a `trie` should look like and what it expects as storage. Memory storage implements Storage which is private and exposed through `RunOnTempTrie`.
- `TransactionStorage` implements `trie.Storage`. It made more sense to move it to `core` since it was the only place it was being used and I wanted to make it obvious that the user of trie would need to provide an implementation of the `Storage` interface.
- The db package looks like the following:
```go
db
|-> bucket.go
|-> bucket_test.go
|-> db.go //defines the db interface 
|-> pebble
  |-> db.go //implements the db interface
  |-> db_test.go
  |-> transaction.go //implements the transaction interface
```
- Move `state` to the `core` package since it made more sense to me.